### DISCUSSION TITLE
Flat arrays

### DISCUSSION THREAD

**Discussion by emil-malmgaard-rasmussen** [react-hook-form/react-hook-form#7770] [General] [answered]
Hello, was wondering if it is going to be possible to have flat arrays, to do something like this:
`
<Controller
    key={field.id}
    name={`categories[${index}]`}
   ...
/>
`
Any time soon?

I keep getting warnings like these:
`"name" prop should be in object shape: name="test[index].name"`
But my endpoint takes an array of objects.

Best Regards!

**Comment by Moshyfawn** [comment, marked as answer]
RHF doesn't support flat arrays ATM.
Ref: [RHF docs on useFieldArray](https://react-hook-form.com/api/usefieldarray#main):
> Name of the field. Important: make sure name is in object shape: name=test.0.name as we don't support flat arrays.

**Reply by emil-malmgaard-rasmussen** [reply]
My question was more directed towards: Is it going to be possible? 😄

**Reply by Moshyfawn** [reply]
Ah, fair enough, my bad :P
@bluebill1049 do we ever wanna support flat arrays? Should we have a feature request created to vote for?

**Reply by emil-malmgaard-rasmussen** [reply]
I guess for now I could use React's useState, append to the state onChange and in handleSubmit(transform) just add the useState to the postRequest. It just isn't as "sexy" as just accessing via data in transform 😅  
But I could read through some old threats that it has been causing some unwanted side-effects, hence my request if it is going to be added in the library in near future 🙂

**Reply by bluebill1049** [reply]
We don't support flat arrays because it's a design limitation on sorting complex data without the unique id provided by react hook form. If you want to use a flat array and its simple enough, you probably get away with something like the below:

```tsx
[...data].map((data, index) => {
  return <input {...register(`data.${index}`)} />
})

const append = (item) => {
  setData((data) => [...data, item])
}

const delete = (index) => unreigseter(`data.${index}`)
```

I haven't tested above, just a theory

**Reply by emil-malmgaard-rasmussen** [reply]
Ahhh okay, well thanks for taking your time answering the question!

I will try it, for now I have fixed it by "cleaning" the data in the transform function like so:
```
let typeSet = new Set();
data.categories.forEach(item => {
    typeSet.add(item.id)
})
data.categories = [...typeSet];
```

**Reply by bluebill1049** [reply]
Let me know how did you go, if it's difficult I will give it a try myself.

**Reply by emil-malmgaard-rasmussen** [reply]
It seems to work, I don't think its perticular bad to be honest, I do like the fact that we have the "transform" function to clean up. The way with new Set(); also removes duplicates. 

Thanks for the answers and this library!

**Reply by bluebill1049** [reply]
Awesome! Would you like to share your solution? It may be helpful to others well

**Reply by emil-malmgaard-rasmussen** [reply]
Yes of course, I actually have another question, I'm not sure what to categorize it as, wether I should open a new "issue".
When I created multiple Controllers with render/as={<CreatableSelect />} (from React Select)
and I supplied the key={field.id} i experienced a bug where, when I had 1 controller and selected a item in the dropdown added a new controller with append({object}) it would clear the previous controller, if I then selected in the 1st controller again, and appended once again, it would stay. this proceed to happen for all the controllers, if you selected, cleared, selected, it would stay there after the second select. This bug wasn't happening after I removed key={field.id}

Thought I would create a Codesandbox but should I open a new ticket? 🙂

**Reply by bluebill1049** [reply]
I think a seperate question is better.

**Reply by pfcodes** [reply]
I understand why a unique `id` is required. But what brought me to this discussion is that in my use case, I need a flat array to store unique IDs. Would be nice if I could do that using `useFieldArray`

**Reply by memben** [reply]
As it took me a few minutes to get the setup in my case right, here is my solution :)

```typescript
// getValues did not work for me
const values = form.watch(`values`) || [];

const append = () => {
  form.setValue(`values`, [...values, ""]);
};

const remove = (index: number) => form.unregister(`values`);
  ```

**Comment by afonsomatos** [comment]
None of the solutions worked for me, here's what worked:

```tsx
export function useFlatArray<TFieldValues extends FieldValues>({
    name,
    control
}: {
    name: FieldArrayPath<TFieldValues>;
    control: Control<TFieldValues>;
}) {
    const fieldArray = useFieldArray({ control, name });
    return {
        ...fieldArray,
        append(value: string) {
            fieldArray.append(
                // @ts-ignore
                new String(value)
            );
        }
    };
}
```

And then, if you're using `zod`, use `z.coerce.string` inside your `z.array`.

**Reply by fekete965** [reply]
That's quite an elegant solution! Thank you! 👍
